src/core/umount.c

```diff
@@ -381,13 +381,24 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+/* This includes remounting readonly, which changes the kernel mount options.
+ * Therefore the list passed to this function is invalidated, and should not be reused. */
+
 static int mount_points_list_umount(MountPoint **head, bool *changed, bool log_error) {
         MountPoint *m, *n;
         int n_failed = 0;
 
         assert(head);
 
         LIST_FOREACH_SAFE(mount_point, m, n, *head) {
+                bool mount_is_readonly = false;
+                const char *word, *state;
+                size_t l;
+
+                FOREACH_WORD_SEPARATOR(word, l, m->options, ",", state) {
```

Comment:
Hmm, we try to avoid using FOREACH_WORD() and friends these days. In fact I think we should work harder to remove its use entirely from our codebase. Use extract_first_word() in a loop instead, it can deal properly with escaping and such, and is generally preferable.

That said, since you are looking into mount options here, please use fstab_test_yes_no_option() or one of the related calls from src/shared/fstab-util.h